### PR TITLE
[ENH] Add Hurdle Likelihood for intermittent demand

### LINKED ISSUE #285: [ENH] Add Hurdle Likelihood for intermittent demand

### PR DESCRIPTION
**Pull request by tingiskhan**
Closes #285 
Adds an intermittent demand forecaster as a pure model.

I really couldn't figure out how to implement this as a fixed effects model as I don't think that my implementation is a good fit for that. Or do you disagree?

### CODE REVIEW

File: src/prophetverse/sktime/intermittent_demand/_truncated_discrete.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+from ._inverse_functions import _inverse_neg_binom, _inverse_poisson
+
+REGISTRY = {
+    dist.Poisson: _inverse_poisson,
+    dist.NegativeBinomial2: _inverse_neg_binom,
+}
+
+
+class TruncatedDiscrete(Distribution):
```
Comment (felipeangelimvieira) [inline src/prophetverse/sktime/intermittent_demand/_truncated_discrete.py:17]: Suggestion: move to the existing "distributions.py" file on the project

Comment (tingiskhan) [reply src/prophetverse/sktime/intermittent_demand/_truncated_discrete.py:17]: Ah, of course, I was looking for one but not well enough it seems... will do!

**Comment by tingiskhan** [comment]
> > I really couldn't figure out how to implement this as a fixed effects model as I don't think that my implementation is a good fit for that. Or do you disagree?
> 
> You mean adding fixed effects for panel data, right? I think we could also add it in a future enhancement.

It was mainly in reference to [your comment](https://github.com/sktime/sktime/pull/8438#issuecomment-3165844272). But I might have misunderstood you?

**Comment by tingiskhan** [comment]
@felipeangelimvieira Alright, ready for review now! I created a dedicated module for distributions to avoid having one big file

**Comment by felipeangelimvieira** [comment]
@tingiskhan made a PR to your fork! Could you take a look when possible please?

**Comment by tingiskhan** [comment]
@felipeangelimvieira I have some changes I haven't pushed yet, will do that when I get home. But would it make sense to merge this and then add AR/RW effects later?

**Comment by felipeangelimvieira** [comment]
> I have some changes I haven't pushed yet, will do that when I get home. But would it make sense to merge this and then add AR/RW effects later?

oh sorry! What changes do you have on your PC? I was making this PR ready to merge. There were some files that I forgot to remove before creating the PR to your fork.

About AR/RW, yes, I think we could add it separatedly

**Comment by felipeangelimvieira** [comment]
One thing that, if you want to, you could also add, is the documentation example for this hurdle model. You could also put your name as the author on that documentation page.

**Comment by tingiskhan** [comment]
Hm, I think I need to take a closer look at the modeling because I can't really reproduce the fit I had in the sktime PR for Joaquin Amat's dataset.

Original:
bild

Current:
bild

I'm using the same pipeline (skipping the fixed effects for Fourier features).

**Comment by tingiskhan** [comment]
@felipeangelimvieira I'm actually not incorporating AR/RW in that modeling scenario as I used MCMC for inference, which caused the model to take too long to run.

But, I do have a hunch as to what's causing the discrepancy: in the original model I regress all of the exogenous variables together with a column of ones. Looking at the code, there should be no difference in theory (at least I think so), but it might be the root cause as that's the main difference between the two implementations. I can confirm that neither `HurdleDistribution` or `TruncatedDiscrete` are the causes as I can replicate the results using the old model and these distributions.

One other thing I noticed is that the inference in the old model relies on `exp` for the demand. Changing to `softplus` results in an extremely poor fit. I can't get `exp` to work in the new version as the loss goes to nan after ~10k iterations, but it doesn't for the old model (using the same inference engine).

**Comment by felipeangelimvieira** [comment]
@tingiskhan  I was wondering what is the difference between the HurdleDistribution + TruncatedDistribution and using ZeroInflatedDistribution from numpyro? Am I missing something?

**Comment by felipeangelimvieira** [comment]
Ok, got good results by re-using for the gate the seasonality component from demand:

image

The results actually seem better than the original? Since the mean is more aligned to the observed value. In the results you shared it seems that the upper bound of the 90% interval was aligning more to the real data

**Comment by tingiskhan** [comment]
I don’t have and more changes, so when you’re happy I’m happy! I’ll open a PR and try to address the AR/RW components

**Comment by felipeangelimvieira** [comment]
For the AR, I have implemented this following your original implementation but not sure if 100% correct:

```python
"""Autoregressive latent effect.

Implements a simple state-space AR(1) latent component optionally combined with
linear regressors provided via ``X``. The latent state is modeled with a
Gaussian transition with (optional) mean reversion parameter ``phi`` in (0, 1)
obtained by applying a sigmoid transform to a Normal prior.

This file exposes:
* ``AREffect``: a ``BaseEffect`` that can be plugged into Prophetverse models
  to provide an additive AR latent component (plus optional exogenous linear
  regressors).

"""
from typing import Any, Dict, Optional

import jax.numpy as jnp
import numpy as np
import numpyro
from numpyro.distributions import Normal, HalfNormal, TransformedDistribution
from numpyro.distributions.transforms import (
    AffineTransform,
    RecursiveLinearTransform,
    SigmoidTransform,
)

from prophetverse.effects.base import BaseEffect

__all__ = ["AREffect"]


class AREffect(BaseEffect):
    """Autoregressive latent effect for additive decomposition.

    Produces an additive latent component following an AR(1) Gaussian process
    with optional linear regressors. The latent state is of length T (or T+H
    during prediction) and returned with shape (T,1).

    Parameters
    ----------
    mean_reverting : bool, default True
        Whether to estimate a mean reversion parameter in (0,1).
    use_regressors : bool, default True
        Whether to include linear regressors on selected ``X`` columns.
    requires_X : bool, default False
        If False, effect can be used with no exogenous regressors.
    """

    _tags = {
        "requires_X": False,  # We can run with or without X
        "filter_indexes_with_forecating_horizon_at_transform": True,
    }

    def __init__(
        self,
        mean_reverting: bool = True,
    ):
        self.mean_reverting = mean_reverting
        super().__init__()

    def _fit(self, y, X, scale=1):
        self.y_index_ = y.index
        return self

    def _transform(self, X, fh):  # noqa: D401
        # Convert to array; allow None

        full_fh = fh.union(self.y_index_)
        # Get positions of fh in full_fh
        oos_idx = full_fh.get_indexer(fh)
        return jnp.array(oos_idx), jnp.arange(len(full_fh))

    def _predict(
        self,
        data: Any,
        predicted_effects: Dict[str, jnp.ndarray],
        *args,
        **kwargs,
    ) -> jnp.ndarray:

        oos_idx, full_fh = data
        sigma = numpyro.sample("sigma", HalfNormal())

        reversion_speed = 1.0
        if self.mean_reverting:
            reversion_speed = numpyro.sample(
                "phi", TransformedDistribution(Normal(scale=1.5), SigmoidTransform())
            )

        transition_matrix = jnp.reshape(reversion_speed, (1, 1))

        eps = Normal().expand((len(full_fh), 1)).to_event(1)

        time_varying_component = numpyro.sample(
            "x:ignore",
            TransformedDistribution(
                eps,
                [
                    AffineTransform(0.0, sigma),
                    RecursiveLinearTransform(transition_matrix=transition_matrix),
                ],
            ),
        )

        return time_varying_component[oos_idx]

    @classmethod
    def get_test_params(cls, parameter_set="default"):
        return [{}]
```

**Comment by tingiskhan** [comment]
Looks good! There _might_ be issues when predicting as I’m not entirely sure whether you can condition only parts of the RV as when using “scan” (which I prefer to use, but “RecursiveLinearTransform” has better inference performance in my experience).

**Comment by tingiskhan** [comment]
I think that the only revision required would be to split the inference and prediction densities as I did. Numpyro >= 0.19 also adds support for initial values for RecursiveLinearTransform, so no need to transform the mean as I did
